### DISCUSSION TITLE
Replace Google fonts with self-hosted fonts

### DISCUSSION THREAD

**Discussion by AlexanderZeilmann** [timlrx/tailwind-nextjs-starter-blog#370] [Show and tell]
## The Problem
This might be relevant to you if you are based in the European Union.
Google fonts is not GDPR compliant (General Data Protection Regulation by the EU) see [here for a court case in german](https://rewis.io/urteile/urteil/lhm-20-01-2022-3-o-1749320/#rd_9), tldr: the website owner had to pay 100€ to a user because Google Fonts was included without asking for permission.
Since asking the user for the permission to include a font is probably not a convenient idea, it is best to self-host the fonts.

Here is a short explanation of how this can be done. Maybe it is useful to someone.

## The Solution
Use [fontsource](https://fontsource.org/). It lets you install fonts like npm packages. Things you need to do:
1. remove the google fonts code from `pages/_document.js`
2. `npm install -save  @fontsource/<font-name>`, repeat for all your fonts
3. Import the fonts in `pages/_app.js`:
```js
import '@fontsource/roboto/400.css'
import '@fontsource/roboto/400-italic.css'
// Add more weights as needed
```
4. Optional: Remove `*.googleapis.com` and `fonts.gstatic.com` from Content Security Policy.

**Variable fonts**

Variable fonts can be imported using
```js
import '@fontsource/roboto-mono/variable.css'
import '@fontsource/roboto-mono/variable-italic.css'
```
and replacing `'Roboto Mono'` by `'Roboto MonoVariable'` in your `tailwind.config.js` (or wherever you are setting the font).

Note, that this change alone does not make your site GDPR compliant, but it is a good start.

## Other Advantages
[See here](https://fontsource.org/docs/introduction)
* Self-hosting brings significant performance gains
* Fonts remain version locked
* Your fonts load offline for PWAs

## Alternatives
* https://google-webfonts-helper.herokuapp.com/fonts Not updated for several years
* https://nextgenthemes.com/google-webfont-downloader/ Only provides woff2 fonts

**Comment by timlrx** [comment]
Fontsource is a really nice approach! It's about as easy to integrate as Google fonts and is self-hosted. Will modify the default template to use this.

**Reply by AlexanderZeilmann** [reply]
I can send a PR if you want 🙂

**Reply by timlrx** [reply]
Sure, happy to accept a PR 😃

**Comment by timlrx** [comment]
Replaced Google Fonts with the self-hosted version in v1.5.0. Thanks @AlexanderZeilmann!

Information on the [performance improvement is in the PR](https://github.com/timlrx/tailwind-nextjs-starter-blog/pull/373#issuecomment-1026883799).

A short guide on the rationale for the change and how to customize the font is also included in the [blog](https://tailwind-nextjs-starter-blog.vercel.app/blog/new-features-in-v1#self-hosted-font-v150).
